### PR TITLE
Fixes for file renaming and manifest features

### PR DESCRIPTION
**Pull request by NikhilCollooru**
```
== NO RELEASE NOTE ==
```

depended by https://github.com/facebookexternal/presto-facebook/pull/1363

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java [hunk 1]
```diff
@@ -346,8 +348,23 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
                             partitionName));
         }
 
-        // Sort FileStatus objects (instead of, e.g., fileStatus.getPath().toString). This matches org.apache.hadoop.hive.ql.metadata.Table.getSortedPaths
-        fileInfos.sort(null);
+        if (fileInfos.get(0).getPath().getName().matches("\\d+")) {
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java:351]: Without file renaming, how did we sort bucket file names before? Do we append 0s in front of file names?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java [hunk 2]
```diff
@@ -112,17 +112,36 @@ public static long getFileSize(Page statisticsPage, int position)
...
+        List<Long> fileSizes = fileWriteInfos.stream().map(FileWriteInfo::getFileSize).filter(Optional::isPresent).map(Optional::get).collect(toImmutableList());
+
+        if (fileSizes.size() < fileNames.size()) {
+            if (fileSizes.size() == 0) {
+                // These files may not have been written by OrcFileWriter. So file sizes not available.
+                return metadata;
+            }
+            throw new PrestoException(
+                    MALFORMED_HIVE_FILE_STATISTICS,
+                    format("During manifest creation for partition= %s, filename count= %s is not equal to filesizes count= %s",
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java:135]: leave the first param to its own line

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java [hunk 3]
```diff
@@ -112,17 +112,36 @@ public static long getFileSize(Page statisticsPage, int position)
...
+        }
+
         // Sort the file infos based on fileName
         fileWriteInfos.sort(Comparator.comparing(info -> Integer.valueOf(info.getWriteFileName())));
 
+        List<String> fileNames = fileWriteInfos.stream().map(FileWriteInfo::getWriteFileName).collect(toImmutableList());
+        List<Long> fileSizes = fileWriteInfos.stream().map(FileWriteInfo::getFileSize).filter(Optional::isPresent).map(Optional::get).collect(toImmutableList());
+
+        if (fileSizes.size() < fileNames.size()) {
+            if (fileSizes.size() == 0) {
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java:129]: `fileSizes.isEmpty()`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java [hunk 4]
```diff
@@ -148,6 +150,18 @@ public static long getFileSize(Page statisticsPage, int position)
         return partitionMetadata.build();
     }
 
+    public static long getManifestSizeInBytes(ConnectorSession session, PartitionUpdate partitionUpdate)
+    {
+        if (isFileRenamingEnabled(session) && partitionUpdate.containsNumberedFileNames()) {
+            List<FileWriteInfo> fileWriteInfos = new ArrayList<>(partitionUpdate.getFileWriteInfos());
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java:156]: why we need a new `ArrayList`?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java [hunk 5]
```diff
@@ -148,6 +150,18 @@ public static long getFileSize(Page statisticsPage, int position)
...
+            return compressFileNames(fileWriteInfos.stream().map(FileWriteInfo::getWriteFileName).collect(toImmutableList())).length()
+                    + compressFileSizes(fileWriteInfos.stream().map(FileWriteInfo::getFileSize).filter(Optional::isPresent).map(Optional::get).collect(toImmutableList())).length();
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java:158]: This is duplicated logic. We can directly get the size info from `extraPartitionMetadata` in `HiveMetadata.finishInsert`. No need to compute it again.

File: presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java (hunk 1 above)
Comment (NikhilCollooru) [reply presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java:351]: Yes correct. We append 0s in front.

**Comment by highker** [comment]
> Yes correct. We append 0s in front.

If we append 0s for bucketed tables, why sorting will be a problem? Or even with bucketed tables + file renaming, we will not do 0 appending anymore?

**Comment by NikhilCollooru** [comment]
> > Yes correct. We append 0s in front.
> 
> If we append 0s for bucketed tables, why sorting will be a problem? Or even with bucketed tables + file renaming, we will not do 0 appending anymore?

In normal scenario, for bucketed tables the names are like "file_prefix + padded_bucket_number" .
https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/HiveWriterFactory.java#L635
The fileName is not efficiently compressable because of the file-prefix.

In file renaming case, we dont append 0 at the beginning of the filename because we dont know how many files are there in the partition . Hence we dont know how many 0s we should pad.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java [hunk 6]
```diff
@@ -149,6 +151,21 @@ public static long getFileSize(Page statisticsPage, int position)
...
+        if (isFileRenamingEnabled(session) && partitionUpdate.containsNumberedFileNames()) {
+            if (parameters.containsKey(FILE_NAMES) && parameters.containsKey(FILE_SIZES)) {
+                return parameters.get(FILE_NAMES).length() + parameters.get(FILE_SIZES).length();
+            }
+            List<FileWriteInfo> fileWriteInfos = partitionUpdate.getFileWriteInfos();
+            return compressFileNames(fileWriteInfos.stream().map(FileWriteInfo::getWriteFileName).collect(toImmutableList())).length()
+                    + compressFileSizes(fileWriteInfos.stream().map(FileWriteInfo::getFileSize).filter(Optional::isPresent).map(Optional::get).collect(toImmutableList())).length();
+        }
+
+        Optional<Page> manifestBlob = createPartitionManifest(partitionUpdate);
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java:165]: `createPartitionManifest` is actually not used anymore. I guess we can remove it

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java [hunk 7]
```diff
@@ -149,6 +151,21 @@ public static long getFileSize(Page statisticsPage, int position)
...
 
+    public static long getManifestSizeInBytes(ConnectorSession session, PartitionUpdate partitionUpdate, Map<String, String> parameters)
+    {
+        if (isFileRenamingEnabled(session) && partitionUpdate.containsNumberedFileNames()) {
+            if (parameters.containsKey(FILE_NAMES) && parameters.containsKey(FILE_SIZES)) {
+                return parameters.get(FILE_NAMES).length() + parameters.get(FILE_SIZES).length();
+            }
+            List<FileWriteInfo> fileWriteInfos = partitionUpdate.getFileWriteInfos();
+            return compressFileNames(fileWriteInfos.stream().map(FileWriteInfo::getWriteFileName).collect(toImmutableList())).length()
+                    + compressFileSizes(fileWriteInfos.stream().map(FileWriteInfo::getFileSize).filter(Optional::isPresent).map(Optional::get).collect(toImmutableList())).length();
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java:162]: We don't need this part, this is duplication with `updatePartitionMetadataWithFileNamesAndSizes`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java [hunk 8]
```diff
@@ -149,6 +151,21 @@ public static long getFileSize(Page statisticsPage, int position)
         return partitionMetadata.build();
     }
 
+    public static long getManifestSizeInBytes(ConnectorSession session, PartitionUpdate partitionUpdate, Map<String, String> parameters)
+    {
+        if (isFileRenamingEnabled(session) && partitionUpdate.containsNumberedFileNames()) {
+            if (parameters.containsKey(FILE_NAMES) && parameters.containsKey(FILE_SIZES)) {
+                return parameters.get(FILE_NAMES).length() + parameters.get(FILE_SIZES).length();
+            }
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java:159]: We just check if the extra params contains `MANIFEST_VERSION`. If yes, we sum up the values' lengths for `FILE_NAMES` and `FILE_SIZES`.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java [hunk 9]
```diff
@@ -149,6 +151,21 @@ public static long getFileSize(Page statisticsPage, int position)
         return partitionMetadata.build();
     }
 
+    public static long getManifestSizeInBytes(ConnectorSession session, PartitionUpdate partitionUpdate, Map<String, String> parameters)
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveManifestUtils.java:154]: return optional long

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 10]
```diff
@@ -1941,10 +1940,8 @@ else if (partitionUpdate.getUpdateMode() == NEW || partitionUpdate.getUpdateMode
                     extraPartitionMetadata = updatePartitionMetadataWithFileNamesAndSizes(partitionUpdate, extraPartitionMetadata);
                 }
 
-                // TODO: Put the manifest blob in partition parameters
                 // Track the manifest blob size
-                Optional<Page> manifestBlob = createPartitionManifest(partitionUpdate);
-                manifestBlob.ifPresent(manifest -> hivePartitionStats.addManifestSizeInBytes(manifest.getRetainedSizeInBytes()));
+                hivePartitionStats.addManifestSizeInBytes(getManifestSizeInBytes(session, partitionUpdate, extraPartitionMetadata));
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:1946]: Only add to the stats if it's not optional empty.

**Review by highker — APPROVED** [review]

**Pull request merged**